### ISSUE TITLE
[DX][RFC] Forcing to https causes a lot of troubles

### ISSUE THREAD

**Issue by Aerendir** [symfony/symfony#27603]
**Symfony version(s) affected**: from 2.8 I think, for sure from 3.4

**Description**  
Forcing to `https` may lead to an infinite loop in some (unclear for me) circumstances.

It is something related to load balancers and trusted proxies.

**How to reproduce**  
1. Force to `https`, [as described here](https://symfony.com/doc/current/security/force_https.html), putting this in `security.yaml`

```yaml
security:
    access_control:
        - { path: ^/, role: IS_AUTHENTICATED_ANONYMOUSLY, requires_channel: https }
```

2. Deploy for example, to Heroku.
3. Access the home page and you will get an error about too many redirects.

**Possible Solution**  
There are two possible solutions to this problem and a required docs improvement:

1. [Manually adding the required `$_SERVER` variables](https://stackoverflow.com/a/34119372/1399706):

```php
// SF4: public/index.php
// SF 2||3: web/app.php

if($_SERVER['HTTP_X_FORWARDED_PROTO'] == 'https')
{
    $_SERVER['HTTPS'] = 'on';
    $_SERVER['SERVER_PORT'] = 443;
}
```

2. [Adding trusted proxies to `framework`](https://stackoverflow.com/a/34120668/1399706):

```yaml
framework:
    trusted_proxies:  [127.0.0.1, ::1]
```

3. Adding to the documentation a note about this common error and the possible solutions
    - [How to Force HTTPS or HTTP for different URLs](https://symfony.com/doc/current/security/force_https.html)
    - Maybe here, too: [How to Force Routes to Always Use HTTPS or HTTP](http://symfony.com/doc/4.0/routing/scheme.html) (I didn't verified if the error happens also forcing `https` using annotations)

**WHAT I'M ASKING FOR**

**A simpler way to understand that there is a problem with the `https` redirection.**

**Maybe a clear exception** in case of excessive number of redirects that tells the developer that it should either implement solution one or solution two.

This takes me 2 days to be solved now that I migrated to SF4 and takes some more days when I implemented `https` forcing using SF3.4 (I forgot about the problem, so I had to discover the solution two times, damn!).

Any opinions about this?

**Comment by stof** [comment]
> Maybe a clear exception in case of excessive number of redirects that tells the developer that it should either implement solution one or solution two.

the thing is, I don't think we can know about this excessive number of redirects, due to the stateless nature of HTTP (the browser can detect such case, as it is the one following redirection).

Among your possible solutions, the first proposal is not a solution. It is a hack trying to partial trust a proxy header without making it a trusted proxy (and if your app is accessible without going through the proxy, it might open a security hole in your app).
When deploying to heroku, the right solution is the second one: as heroku runs a proxy in front of your app (their load balancing router), you need to configure the proxy as trusted. and Heroku even documents it: https://devcenter.heroku.com/articles/getting-started-with-symfony#trusting-the-load-balancer

**Comment by Aerendir** [comment]
@stof , you are right: the first solution is a hack and the preferred way of solving the problem is adding the trusted proxies.

Anyway, the problem is that, before understanding what is really going under the hood and from where it comes the redirection, you have to spend a lot of time.

I don't know if it is possible to intercept the loop: I cannot come up with a solution but maybe someone of you with more experience can :)

Anyway, if no solution is possible, the best approach would be to add to the docs a note about this problem that is known and intended, maybe also a link to the Heroku documentation or to the `trusted_proxies`... I don't know.

The only thing I know is that to discover the source of the loop I spent really a lot of time as this is something you set once and then forget.

Just as a side note, thank you for the link to the Heroku documentation: when I started with Heroku it didn't exist and I didn't find it searching for the issue on Google.

**Comment by colinodell** [comment]
What about something like this in `ChannelListener::handle()`? (untested)

```diff
     public function handle(GetResponseEvent $event)
     {
         $request = $event->getRequest();
 
         list(, $channel) = $this->map->getPatterns($request);
 
         if ('https' === $channel && !$request->isSecure()) {
             if (null !== $this->logger) {
                 $this->logger->info('Redirecting to HTTPS.');
+                if ('https' === $request->headers->get('X-Forwarded-Proto')) {
+                    $this->logger->info('Possible redirect loop');
+                }
             }

             $response = $this->authenticationEntryPoint->start($request);

             $event->setResponse($response);

             return;
         }
```

That would at least give you a hint in the logs.

Of course this is also one of those problems that once you (finally) figure it out the first time, it's easy to spot and fix again in the future.

**Comment by Aerendir** [comment]
@colinodell , yes, but this is also something that yu set once and then forget about it.

If you only manage one application (like me), you forget about it.

I spotted the problem about 1 year ago the first time and now, upgrading to SF4 I had to spot it again as I completely forgot I used the hack...

More, as I didn't know anything about this problem, I used the hack, so my app was potentially vulnerable without I knowing nothing about.

For me a good approach would be to, other than logging the message like you shown, also to add a not in the documentation about the potential loop where is explained how to force `https`.
